ultravox/data/datasets.py

```diff
@@ -947,6 +949,67 @@ def _get_sample(self, row) -> VoiceSample:
         )
 
 
+class GenericVoiceDataset(VoiceDataset):
+    def __init__(
+        self, args: VoiceDatasetArgs, config: config_base.DataDictConfig
+    ) -> None:
+        super().__init__(args)
+
+        dataset = datasets.concatenate_datasets(
+            [
+                self._load_audio_dataset(
+                    config.path,
+                    name=config.name,
+                    split=s,
+                    streaming=config.streaming,
+                    shuffle=False,
+                )
+                for s in config.splits
+            ]
+        )
+
+        if self._args.shuffle:
+            dataset = dataset.shuffle(seed=self._args.shuffle_seed)
+
+        if config.num_samples:
+            dataset = dataset.select(range(config.num_samples))
```

Comment:
There could be situations where you want to only use certain number of samples from a dataset for training, for example to achieve some balance between different datasets.

Reply:
Hmm, I'm not sure how that would actually work in practice with the current code-base unless you come up with an alternative for `InterleaveDataset`.

If you set `repeat=False`, this can have an effect, but `repeat=False` in its current form is not very desirable because the composition of data changes in later batches in ways that are very hard to reason about.

All of that aside, you again need a comment for not working with MDS datasets and you also need an if condition to handle the streaming and non-streaming cases (i.e. `select` vs `take`).

That reminds me, are you sure the code works with both streaming and non-streaming datasets? IIRC `VoiceDataset` assumes that the dataset in `_init_dataset` is iterable, but I might be wrong.

Reply:
These are good points, need to look into the details.

Reply:
I think I'm actually wrong about `_init_dataset` expecting an `IterableDataset`. I just haven't tested it in practice. We only assume that it can be iterated on which both `Dataset` and `IterableDataset` probably support :?

Reply:
tbh I haven't given it much thought. I simply included it because the use of num_samples in TrainConfig suggests that it could be a nice feature for individual datasets too.